File: solutions/3097. Shortest Subarray With OR at Least K II/3097.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
  def update(self, bits: List[int], x: int, change: int) -> None:
    for i in range(32):
      if (x >> i) & 1 != 0:
        bits[i] += change

  def bitsToNum(self, bits: List[int]) -> int:
    res = 0

    for i in range(32):
      if bits[i] != 0:
        res |= 1 << i

    return res        

  def minimumSubarrayLength(self, nums: List[int], k: int) -> int:
    ans = len(nums) + 1
    bits = [0] * 32

    l = 0
    for r in range(len(nums)):
      self.update(bits, nums[r], 1)

      while l <= r and self.bitsToNum(bits) >= k:
        ans = min(ans, r - l + 1)
        self.update(bits, nums[l], -1)
        l += 1

    return ans if ans != len(nums) + 1 else -1

File: solutions/3097. Shortest Subarray With OR at Least K II/3097.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  int minimumSubarrayLength(const vector<int>& nums, int k) {
    int ans = nums.size() + 1;
    vector<int> bits(32);
    for (int l = 0, r = 0; r < nums.size(); ++r) {
      update(bits, nums[r], 1);
      while (l <= r && bitsToNum(bits) >= k) {
        ans = min(ans, r - l + 1);
        update(bits, nums[l++], -1);
      }
    }
    return ans != nums.size() + 1 ? ans : -1;
  }
 private:
  void update(vector<int>& bits, int x, int change) {
    for (int i = 0; i < 32; ++i)
      if ((x >> i) & 1)
        bits[i] += change;
  }
    
  int bitsToNum(const vector<int>& bits) {
    int res = 0;

    for (int i = 0; i < 32; ++i)
      if (bits[i])
        res |= 1 << i;

    return res;
  }    
};